cloudinit/ssh_util.py

```diff
@@ -321,17 +321,27 @@ def check_create_path(username, filename, strictmodes):
         home_folder = os.path.dirname(user_pwent.pw_dir)
         for directory in directories:
             parent_folder += "/" + directory
-            if home_folder.startswith(parent_folder):
+            if (home_folder.startswith(parent_folder) or
+                    parent_folder == user_pwent.pw_dir):
                 continue
 
             if not os.path.isdir(parent_folder):
                 # directory does not exist, and permission so far are good:
                 # create the directory, and make it accessible by everyone
                 # but owned by root, as it might be used by many users.
                 with util.SeLinuxGuard(parent_folder):
-                    os.makedirs(parent_folder, mode=0o755, exist_ok=True)
-                    util.chownbyid(parent_folder, root_pwent.pw_uid,
-                                   root_pwent.pw_gid)
+                    mode = 0o755
+                    uid = root_pwent.pw_uid
+                    gid = root_pwent.pw_gid
+                    if parent_folder.startswith(user_pwent.pw_dir):
+                        mode = 0o700
+                        uid = user_pwent.pw_uid
+                        gid = user_pwent.pw_gid
+                    try:
+                        os.mkdir(parent_folder, mode=mode)
+                    except FileExistsError:
```

Comment:
So with this change 
- `/`, `/home/` and `/home/<user>` are skipped and not created in any case. 
- then, we can have either a path that is not in the user home folder, and we chmod it to root ownership, or inside the user folder, and we chmod it as user-owned.

Only one question: what if `/home/` or `/home/<user>` do not exist? `os.mkdir` should launch a `FileNotFoundError` I think, we should handle also this case right?

Reply:
I think I agree with @paride in https://bugs.launchpad.net/cloud-init/+bug/1940233/comments/4 when he says it's not the responsibility of this module to create a user's home directory.

Reply:
A couple of thoughts about this:
 * The `authorized_keys` file is always created user-owned (chmod 600), even if its path is outside the user's `$HOME` directory. I think this is good, so users are still be able to manage their own `authorized_keys`.
 * Given that we've got bitten by this, I think it we should test two more scenarios:
   * AuthorizedKeysFile outside of `$HOME`, e.g. in `/etc/ssh/authorized_keys/%u/`
   * AuthorizedKeysFile deeper in `$HOME`, e.g. `%h/secret/ssh/`

Reply:
@TheRealFalcon ok, seems reasonable
@paride regarding the right permission w.r.t. authorized_keys files, look [here](https://github.com/canonical/cloud-init/pull/948#issuecomment-891332457). The mistake in this code is that I did not think about the permissions that other ssh tools need, in this case `ssh-keygen`. Because if `.ssh` has permission 755 and it's owned by root, it should work w.r.t. AuthorizedKeysFile. 

After a quick test I see that the mode doesn't matter, the ownership is the problem (which makes sense because with root owner and 755 `ssh-keygen` on behalf of the user cannot write in the folder).
 
Regarding the first scenario, it is covered in [this](https://github.com/canonical/cloud-init/blob/main/tests/unittests/test_sshutil.py#L1177) and [this](https://github.com/canonical/cloud-init/blob/main/tests/unittests/test_sshutil.py#L1108) test.

Reply:
Added the requested integration tests. There's some overlap with the unit tests (and the unit tests are much more comprehensive), but better to have a basic integration check of the other use cases too.

Reply:
Sounds good to me, thank you!

Reply:
LGTM, thanks!